ext/sqlite3/database.c

```diff
@@ -251,6 +251,51 @@ busy_handler(int argc, VALUE *argv, VALUE self)
     return self;
 }
 
+static int
+rb_sqlite3_progress_handler(void *ctx)
+{
+  VALUE self = (VALUE)(ctx);
+  VALUE handle = rb_iv_get(self, "@progress_handler");
+  VALUE result = rb_funcall(handle, rb_intern("call"), 0);
+
+  if (Qfalse == result) return 1;
+
+  return 0;
+}
+
+/* call-seq:
+ *    progress_handler([n]) { ... }
+ *    progress_handler([n,] Class.new { def call; end }.new)
+ *
+ * Register a progress handler with this database instance.
+ * This handler will be invoked periodically during a long-running query or operation.
+ * If the handler returns +false+, the operation will be interrupted; otherwise, it continues.
+ * The parameter 'n' specifies the number of SQLite virtual machine instructions between invocations.
+ * If 'n' is not provided, the default value is 1.
+ */
+static VALUE
+progress_handler(int argc, VALUE *argv, VALUE self)
+{
+  sqlite3RubyPtr ctx;
+  VALUE block, n_value;
+
+  TypedData_Get_Struct(self, sqlite3Ruby, &database_type, ctx);
+  REQUIRE_OPEN_DB(ctx);
+
+  rb_scan_args(argc, argv, "02", &n_value, &block);
+
+  int n = NIL_P(n_value) ? 1 : NUM2INT(n_value);
+  if(NIL_P(block) && rb_block_given_p()) block = rb_block_proc();
+
+  rb_iv_set(self, "@progress_handler", block);
+
+  sqlite3_progress_handler(
+      ctx->db, n, NIL_P(block) ? NULL : rb_sqlite3_progress_handler, (void *)self);
```

Comment:
You can't store a `VALUE` in C-land like this without it being marked, as the object might be moved by the GC, and end up pointing to either garbage or another object.

You can likely cause a crash by calling [`GC.verify_compaction_references`](https://github.com/msgpack/msgpack-ruby/blob/e52d18bacc5a69bd0c823e8daf2792b187a6be46/spec/spec_helper.rb#L21-L25) from inside the callback.

Instead you'd have to store that reference inside the `sqlite3Ruby` struct, and have it marked with `gc_mark` so it's pinned.

Reply:
I _think_ I understand what I should do, but to confirm: would this not also be a problem for the `busy_handler` implementation, from which this line of code is nearly identically copied (I just added the `n` argument). I do not see that object stored inside the `sqlite3Ruby` struct. If the `busy_handler` implementation is safe, could you explain to me the mechanisms that make it safe so that I can mirror those mechanisms for `progress_handler`?

Reply:
No you are right, the busy handler has the same issue, which I hadn't noticed, it also need to be fixed. To be fair however the busy handler was added 15 years ago, when compaction wasn't a concern.

Reply:
Not a judgement of the code, just looking to find pre-existing solutions to mirror if possible. But it sounds like this PR will need to be the first implementation of compaction-resilient handling of such code, which we can then bring back to the `busy_handler`. I'll work on an initial attempt.

Reply:
This is definitely stretching me beyond my current very basic understanding of C and writing C extensions for Ruby.

I just pushed 35a003a which doesn't compile. On my Mac, I get this error and warning:
```
../../../../ext/sqlite3/database.c:49:5: warning: variable 'ctx' is uninitialized when used here [-Wuninitialized]
    ctx->progress_handler = Qnil;
    ^~~
../../../../ext/sqlite3/database.c:48:23: note: initialize the variable 'ctx' to silence this warning
    sqlite3RubyPtr ctx;
                      ^
                       = NULL
../../../../ext/sqlite3/database.c:262:32: error: member reference base type 'void' is not a structure or union
  VALUE result = rb_funcall(ctx->progress_handler, rb_intern("call"), 0);
                            ~~~^ ~~~~~~~~~~~~~~~~
1 warning and 1 error generated.
```

I also don't know how or where to call `rb_gc_mark` with the value object. My hope is that you can help me polish this initial, broken C code @byroot.

Reply:
I can open a PR to fix the busy handler later today if you want.

Reply:
Here you go: https://github.com/sparklemotion/sqlite3-ruby/pull/466

To understand that PR I recommend Peter excellent blog series: https://blog.peterzhu.ca/ruby-c-ext-part-7/

Reply:
Mirrored solution in https://github.com/sparklemotion/sqlite3-ruby/pull/466